TextIndicator does not move when page layout changes.

megangardner opened the pull request:
#### 7f55413e7a0da809f84ed89ba62d2ed4bcf941db
<pre>
TextIndicator does not move when page layout changes.
<a href="https://bugs.webkit.org/show_bug.cgi?id=290430">https://bugs.webkit.org/show_bug.cgi?id=290430</a>
<a href="https://rdar.apple.com/147885631">rdar://147885631</a>

Reviewed by Wenson Hsieh.

We lacked a way to update the text indicator when the
page layout changed, so it was possible for the text indicator
to become disconnected from the text that is was indicating.
This code adds a way to update the text indicator, but it
only fixes the problems on mac, and when it is related to page
layout. We also have this issue on iOS and in relation to
scrolling, and those will be fixed separately.

Most of this is plumbing code to allow the text indicator to update,
and we just remake the indicator in the window and the layer
when an update is requested. This could be improved to allow
the animation to be completed when the update happens by just
updating the text on the layer instead of remaking the whole thing.

* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyChromeClient::updateTextIndicator const):
* Source/WebCore/loader/EmptyClients.h:
* Source/WebCore/page/ChromeClient.h:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::reset):
(WebCore::LocalFrameView::scrollToFragment):
(WebCore::LocalFrameView::textFragmentIndicatorTimerFired):
(WebCore::LocalFrameView::cancelScheduledTextFragmentIndicatorTimer):
(WebCore::LocalFrameView::scheduleScrollToAnchorAndTextFragment):
(WebCore::LocalFrameView::scrollToTextFragmentRange):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/cocoa/WebTextIndicatorLayer.h:
* Source/WebCore/page/cocoa/WebTextIndicatorLayer.mm:
(-[WebTextIndicatorLayer updateWithFrame:textIndicator:margin:offset:updatingIndicator:]):
(-[WebTextIndicatorLayer initWithFrame:textIndicator:margin:offset:]):
* Source/WebCore/page/mac/TextIndicatorWindow.h:
* Source/WebCore/page/mac/TextIndicatorWindow.mm:
(WebCore::TextIndicatorWindow::setTextIndicator):
(WebCore::TextIndicatorWindow::updateTextIndicator):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateTextIndicatorFromFrame):
(WebKit::WebPageProxy::updateTextIndicator):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::updateTextIndicator):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView setUpTextIndicator:]):
(-[WKContentView updateTextIndicator:]):
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::updateTextIndicator):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::setTextIndicator):
(WebKit::WebViewImpl::updateTextIndicator):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::updateTextIndicator const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updateTextIndicator):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::updateTextIndicator const):

Canonical link: <a href="https://commits.webkit.org/292728@main">https://commits.webkit.org/292728@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cb06bcf621d6abcfab5fa6e6a241ef37d6a45774

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96864 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16478 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6699 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101937 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47385 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98909 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16774 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24924 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73810 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31018 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99867 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12631 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/87657 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54144 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12392 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5478 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46712 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82486 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5559 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103961 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23932 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17464 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82861 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24183 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83731 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82253 "Found 3 new API test failures: /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print-errors, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/close-after-print (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20682 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26907 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4487 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17431 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23894 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29049 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23553 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27033 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25294 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -3827,6 +3794,50 @@ void LocalFrameView::scrollToTextFragmentRange()
+    if (m_delayedTextFragmentIndicatorTimer.isActive())
+        return;
+
+    auto textIndicator = TextIndicator::createWithRange(range, { TextIndicatorOption::DoNotClipToVisibleRect }, WebCore::TextIndicatorPresentationTransition::Bounce);
```
```suggestion
    RefPtr textIndicator = TextIndicator::createWithRange(range, { TextIndicatorOption::DoNotClipToVisibleRect }, WebCore::TextIndicatorPresentationTransition::Bounce);
```

aprotyas commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -3827,6 +3794,50 @@ void LocalFrameView::scrollToTextFragmentRange()
+
+    auto textIndicator = TextIndicator::createWithRange(range, { TextIndicatorOption::DoNotClipToVisibleRect }, WebCore::TextIndicatorPresentationTransition::Bounce);
+
+    auto* page = m_frame->page();
```
```suggestion
    RefPtr page = m_frame->page();
```

aprotyas commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -3827,6 +3794,50 @@ void LocalFrameView::scrollToTextFragmentRange()
+
+        auto textRects = RenderObject::absoluteTextRects(range);
+
+        constexpr auto hitType = OptionSet { HitTestRequest::Type::ReadOnly, HitTestRequest::Type::Active, HitTestRequest::Type::AllowVisibleChildFrameContentOnly };
```
You can skip `auto` if you use direct initialization
```suggestion
        static constexpr OptionSet hitType { HitTestRequest::Type::ReadOnly, HitTestRequest::Type::Active, HitTestRequest::Type::AllowVisibleChildFrameContentOnly };
```

Also, optional, but consider making it `static`.

aprotyas commented on Source/WebCore/page/cocoa/WebTextIndicatorLayer.h:
```diff
@@ -40,7 +40,12 @@ WEBCORE_EXPORT @interface WebTextIndicatorLayer : CALayer {
 }
 
-- (instancetype)initWithFrame:(CGRect)frame textIndicator:(WebCore::TextIndicator&)textIndicator margin:(CGSize)margin offset:(CGPoint)offset;
+enum UpdatingIndicator : BOOL { No, Yes };
```
Scoped enums!

```suggestion
enum class UpdatingIndicator : BOOL { No, Yes };
```

aprotyas commented on Source/WebCore/page/cocoa/WebTextIndicatorLayer.h:
```diff
@@ -40,7 +40,12 @@ WEBCORE_EXPORT @interface WebTextIndicatorLayer : CALayer {
 
-- (instancetype)initWithFrame:(CGRect)frame textIndicator:(WebCore::TextIndicator&)textIndicator margin:(CGSize)margin offset:(CGPoint)offset;
+enum UpdatingIndicator : BOOL { No, Yes };
+
```
nit: Extra line here
```suggestion
```

aprotyas commented on Source/WebCore/page/cocoa/WebTextIndicatorLayer.mm:
```diff
@@ -91,18 +91,16 @@ static bool indicatorWantsFadeIn(const WebCore::TextIndicator& indicator)
-    
     self.anchorPoint = CGPointZero;
     self.frame = frame;
-    self.name = @"WebTextIndicatorLayer";
```
Was the name misleading or unhelpful?

aprotyas replied:
Nevermind, just scrolled down!

pxlcoder commented on Source/WebCore/page/cocoa/WebTextIndicatorLayer.h:
```diff
@@ -40,7 +40,12 @@ WEBCORE_EXPORT @interface WebTextIndicatorLayer : CALayer {
 }
 
-- (instancetype)initWithFrame:(CGRect)frame textIndicator:(WebCore::TextIndicator&)textIndicator margin:(CGSize)margin offset:(CGPoint)offset;
+enum UpdatingIndicator : BOOL { No, Yes };
```
I don't think we need an enum for this, since it's only used in Obj-C the method name already provides the description: `updatingIndicator:`.

megangardner replied:
@smfr likes them to be there all the time.

whsieh replied:
In case we _do_ decide to keep this, it should be `enum class UpdatingIndicator : bool { No, Yes };`, preferably inside of a `namespace WebKit` scope (so it won't collide with any other global symbols from the SDK).

FWIW, I don't really have a strong preference here, but I do lean on the side that it's not necessary in this case. The reason strongly typed enums are generally useful in C++ code is that they avoid lists of parameters/initialization arguments that are like `true, false, true, false` (where it's easy to accidentally confuse two values). Using strongly typed enums turns this into a compiler error.

For ObjC methods like this, all parameters always come with labels, which avoids this issue.

rr-codes replied:
I strongly agree with Aditya too

megangardner replied:
@aprotyas added the class in another comment :) I'll add namespace WebKit as well

pxlcoder replied:
@smfr really?? 😛

smfr replied:
I think the "use enums always" only applies to C++. Fine to just use BOOLs in ObjC.

megangardner replied:
Ok, I'll take it out.

aprotyas commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -9896,6 +9896,33 @@ void WebPageProxy::setTextIndicator(const TextIndicatorData& indicatorData, uint
+    });
+}
+
+void WebPageProxy::updateTextIndicator(const TextIndicatorData& indicatorData)
```
Is it possible to make this accept an rvalue ref? That way you could move your text indicator data along everywhere

megangardner replied:
I'll look into this as a follow up

rr-codes commented on Source/WebCore/page/cocoa/WebTextIndicatorLayer.mm:
```diff
@@ -201,6 +201,17 @@ - (instancetype)initWithFrame:(CGRect)frame textIndicator:(WebCore::TextIndicato
+
+    self.name = @"WebTextIndicatorLayer";
+
+    [self updateWithFrame:frame textIndicator:textIndicator  margin:margin offset:offset updatingIndicator:UpdatingIndicator::No];
```
```suggestion
    [self updateWithFrame:frame textIndicator:textIndicator margin:margin offset:offset updatingIndicator:UpdatingIndicator::No];
```

whsieh commented on Source/WebCore/page/mac/TextIndicatorWindow.mm:
```diff
@@ -134,13 +136,41 @@ - (BOOL)isFlipped
         m_temporaryTextIndicatorTimer.startOneShot(WebCore::timeBeforeFadeStarts);
 }
 
+void TextIndicatorWindow::updateTextIndicator(Ref<TextIndicator> textIndicator, CGRect textBoundingRectInScreenCoordinates)
```
```suggestion
void TextIndicatorWindow::updateTextIndicator(Ref<TextIndicator>&& textIndicator, CGRect textBoundingRectInScreenCoordinates)
```

whsieh commented on Source/WebCore/page/mac/TextIndicatorWindow.mm:
```diff
@@ -134,13 +136,41 @@ - (BOOL)isFlipped
+    CGFloat horizontalMargin = dropShadowBlurRadius * 2 + TextIndicator::defaultHorizontalMargin;
+    CGFloat verticalMargin = dropShadowBlurRadius * 2 + TextIndicator::defaultVerticalMargin;
+
+    if (m_textIndicator->wantsBounce()) {
```
To fix the smartpointer error:

```suggestion
    bool wantsBounce = textIndicator->wantsBounce();
    if (m_textIndicator != textIndicator.ptr())
        m_textIndicator = WTFMove(textIndicator);

    CGFloat horizontalMargin = dropShadowBlurRadius * 2 + TextIndicator::defaultHorizontalMargin;
    CGFloat verticalMargin = dropShadowBlurRadius * 2 + TextIndicator::defaultVerticalMargin;

    if (wantsBounce) {
```

megangardner replied:
Well, this isn't how I fixed it last time, but whatever.

whsieh commented on Source/WebCore/page/cocoa/WebTextIndicatorLayer.mm:
```diff
@@ -147,7 +145,8 @@ - (instancetype)initWithFrame:(CGRect)frame textIndicator:(WebCore::TextIndicato
         [bounceLayer setDelegate:[WebActionDisablingCALayerDelegate shared]];
         [bounceLayer setFrame:bounceLayerRect];
-        [bounceLayer setOpacity:0];
+        if (updatingIndicator == NO)
```
Per [style guidelines](https://webkit.org/code-style-guidelines/#zero-comparison):

```suggestion
        if (!updatingIndicator)
```

rr-codes replied:
```suggestion
        if (!updatingIndicator)
```

whsieh commented on Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
```diff
@@ -580,6 +580,11 @@ - (BOOL)_hostsLayersInWindowServer;
     m_impl->setTextIndicator(textIndicator.get(), lifetime);
 }
 
+void PageClientImpl::updateTextIndicator(Ref<TextIndicator> textIndicator)
```
As Abrar mentioned in a previous comment, this (and nearby code) should either pass `Ref<TextIndicator>&&` or `const TextIndicator&`. Maybe we can clean that up separately, though.

megangardner replied:
Yes, I said I’d take care of that separately, there’s going to be serval follow on patches to this problem, it could also go in one of those.

whsieh commented on Source/WebKit/UIProcess/mac/WebViewImpl.h:
```diff
@@ -474,6 +474,7 @@ class WebViewImpl final : public CanMakeWeakPtr<WebViewImpl>, public CanMakeChec
     void preferencesDidChange();
 
     void setTextIndicator(WebCore::TextIndicator&, WebCore::TextIndicatorLifetime = WebCore::TextIndicatorLifetime::Permanent);
+    void updateTextIndicator(WebCore::TextIndicator&);
```
Is there a reason this is a non-const ref? (I think it's just plumbed into the text indicator window as `Ref<TextIndicator>&&`, so I think that `const WebCore::TextIndicator&` should be okay.

megangardner replied:
Because it’s following setTextIndicator above and that is how it is passed

megangardner committed:
TextIndicator does not move when page layout changes.
https://bugs.webkit.org/show_bug.cgi?id=290430
rdar://147885631

Reviewed by Wenson Hsieh.

We lacked a way to update the text indicator when the
page layout changed, so it was possible for the text indicator
to become disconnected from the text that is was indicating.
This code adds a way to update the text indicator, but it
only fixes the problems on mac, and when it is related to page
layout. We also have this issue on iOS and in relation to
scrolling, and those will be fixed separately.

Most of this is plumbing code to allow the text indicator to update,
and we just remake the indicator in the window and the layer
when an update is requested. This could be improved to allow
the animation to be completed when the update happens by just
updating the text on the layer instead of remaking the whole thing.

* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyChromeClient::updateTextIndicator const):
* Source/WebCore/loader/EmptyClients.h:
* Source/WebCore/page/ChromeClient.h:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::reset):
(WebCore::LocalFrameView::scrollToFragment):
(WebCore::LocalFrameView::textFragmentIndicatorTimerFired):
(WebCore::LocalFrameView::cancelScheduledTextFragmentIndicatorTimer):
(WebCore::LocalFrameView::scheduleScrollToAnchorAndTextFragment):
(WebCore::LocalFrameView::scrollToTextFragmentRange):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/cocoa/WebTextIndicatorLayer.h:
* Source/WebCore/page/cocoa/WebTextIndicatorLayer.mm:
(-[WebTextIndicatorLayer updateWithFrame:textIndicator:margin:offset:updatingIndicator:]):
(-[WebTextIndicatorLayer initWithFrame:textIndicator:margin:offset:]):
* Source/WebCore/page/mac/TextIndicatorWindow.h:
* Source/WebCore/page/mac/TextIndicatorWindow.mm:
(WebCore::TextIndicatorWindow::setTextIndicator):
(WebCore::TextIndicatorWindow::updateTextIndicator):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateTextIndicatorFromFrame):
(WebKit::WebPageProxy::updateTextIndicator):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::updateTextIndicator):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView setUpTextIndicator:]):
(-[WKContentView updateTextIndicator:]):
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::updateTextIndicator):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::setTextIndicator):
(WebKit::WebViewImpl::updateTextIndicator):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::updateTextIndicator const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updateTextIndicator):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::updateTextIndicator const):

Canonical link: https://commits.webkit.org/292728@main
